pyvisa-py/gpib.py

```diff
@@ -38,6 +38,46 @@ def _inner(self):
     raise
 
 
+class GPIBEvent(object):
```

Comment:
As I mentionned in my review, I am not sure how to expose this since pyvisa does not expose events. We should perhaps think about how to expose this in pyvisa and how to make it compatible with pyvisa-py.

Reply:
`GPIBEvent` is not a user-facing class, it is an implementation detail: `PyVisaLibrary` stores `GPIBEvent` objects in the `PyVisaLibrary.events` dict and passes `any get_attribute()` and `close()` calls down to them. Such "attribute query workers" are required pieces of the puzzle to fully conform to pyvisa's handle-based backend interface.

I agree that pyvisa needs a more pythonic way to expose events, and I think it is actually quite straightforward to get there. Currently, events _are_ actually exposed to the user as raw VISA event contexts/handles. This is not very pythonic but is nevertheless there as a public and documented API. Our user can get hold of an event handle in two main ways:

1. `visalib` calls the installed event handlers and passes the event handle as a parameter,
2. `Resource.wait_on_event()` returns a `WaitResponse` object which exposes a member called `context`.

In both cases, the user can find out more about the event by calling `visalib.get_attribute(handle, attribute)` or close the event using `visalib.close(handle)`.

Instead of low-level handles/contexts, pyvisa should give users some kind of an event object that can be queried and knows how to clean up when it is done. The `WaitResponse` class already fits the bill nicely: all it needs to become user-friendly are `get_attribute()` and `close()` methods and renaming it eg. `Event`. The remaining step is to wrap user-provided event handlers so that they are passed `Event` objects when called instead of raw event handles.

Reply:
Indeed this sounds like a plan.

For the work at hand, even if we do not implement event support right away, we can probably set up a generic class that could be used by all resources rather than an explicitly GPIB restricted class, no ?

Reply:
Right. Looks pretty generic actually. If any session needs to compute event attributes on the fly, they can override get_attributes(). Right now I'm struggling with bikeshedding :) I think "Event" should only be the name of the user-facing pyvisa class.